libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/processor/MediatorReplayService.kt

```diff
@@ -49,6 +51,29 @@ class MediatorReplayService @Activate constructor(
         return mediatorOutputs
     }
 
+    /**
+     * Compare the [inputRecord] to the existing [mediatorState] to see if it is a replayed record or not.
+     * If it is replay then return all the outputs from the [mediatorState] as [MediatorMessage]s.
+     * @param inputRecord Record to check whether it is a replay or not
+     * @param mediatorState The mediator state to check the [inputRecord] against
+     * @return Null if it is not a replayed event, if it is a replay event,
+     * a list of mediator messages are returned associated with the [inputRecord].
+     */
+    fun <K : Any, V : Any> getReplayEvents(inputRecord: Record<K, V>, mediatorState: MediatorState): List<MediatorMessage<Any>>? {
+        val savedOutputs = mediatorState.outputEvents
+        val inputHash = getInputHash(inputRecord).array()
+
+        savedOutputs.forEach { mediatorReplayOutputEvents ->
+            if (inputHash.contentEquals(mediatorReplayOutputEvents.inputEventHash.array())) {
+                return mediatorReplayOutputEvents.outputEvents.map { outputEvent ->
+                    outputEvent.toMediatorMessage()
+                }
+            }
+        }
```

Comment:
We should leverage our knowledge of the underlying data and search from the end of our list, given that our `savedOutputs` are in chronological order and replayed events are more likely to be at the end. This way the lookup time doesn't increase with the size of the list (provided there is a matching value there).
```suggestion
        return mediatorState.outputEvents
            .findLast { inputHash.contentEquals(it.inputEventHash.array()) }
            ?.outputEvents
            ?.map { it.toMediatorMessage() }
```

Reply:
ah yes, i forgot you suggested this, thanks!